README.md
![Pub Version](https://img.shields.io/pub/v/peer5?style=for-the-badge)

# Peer5 Flutter

Flutter plugin to integrate with the Peer5 native SDKs on Android and iOS.

## Getting Started

You will need some steps to install and configure this plugin properly.

### Step 1: Obtain an API KEY

Register at https://app.peer5.com/register and retrieve the `CustomerID` (aka `API KEY`) in the `Account` page.

### Step 2: Install the plugin

In your `pubspec.yaml` file, add the `peer5` dependency. [Read this page for more information and version number.](https://pub.dev/packages/peer5#-installing-tab-)

Don't forget to run `pod install` inside the `ios/` directory.

### Step 3: Configure iOS and Android

### On iOS

Add your `CustomerID` and, in order to allow the loading of distributed content via the Peer5 proxy, enable loading data from HTTP in your app by opening your `Info.plist` file as source code and adding the following values below the `</dict>` tag:

```xml
<key>Peer5ApiKey</key>
<string>Your-CustomerID-Goes-Here</string>
<key>NSAppTransportSecurity</key>
<dict>
    <key>NSAllowsArbitraryLoads</key>
    <true/>
</dict>
```

#### On Android

Add this metadata tag to your `android/app/src/main/AndroidManifest.xml` inside the `<application>` tag after all activities:

```xml
<meta-data
    android:name="com.peer5.ApiKey"
    android:value="__YOUR_PEER5_TOKEN__" />
```

Starting Android 9 HTTP traffic is disabled by default. read more about this change [here](https://developer.android.com/training/articles/security-config#CleartextTrafficPermitted).

Without allowing HTTP for `127.0.0.1` this error might show up when trying to play: "Cleartext HTTP traffic to 127.0.0.1 not permitted."

Create file in `android/app/src/main/res/xml/network_security_config.xml` with content:

```xml
<?xml version="1.0" encoding="utf-8"?>
<network-security-config>
    <domain-config cleartextTrafficPermitted="true">
        <domain includeSubdomains="true">127.0.0.1</domain>
    </domain-config>
</network-security-config>
```

In `android/app/src/main/AndroidManifest.xml` add the `android:networkSecurityConfig="@xml/network_security_config"` attribute to the `<application>` tag:

```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest>
    <uses-permission android:name="android.permission.INTERNET" />
    <application
        android:networkSecurityConfig="@xml/network_security_config">

    </application>
</manifest>
```

### Step 4: Usage

On iOS, you will need to initialize the local server by calling `Peer5.init();`. The instance should be bound to the application's lifetime and be initialized just once.

You can initialize anytime, but I recommend doing it at an early stage of the app initialization. In the `initState()` of the first widget, or even on the `main()` function (Just remember to add the `WidgetsFlutterBinding.ensureInitialized();` in this case).

On Android, the Peer5 native SDK already starts the server for you. Calling `Peer5.init();` on Android is safe and won't have any effect, so you don't need to add any conditional for it.

When initializing the `VideoPlayerController` instance, you will pass the proxied url that Peer5 will provide by calling `Peer5.getStreamUrl(url)` first.

```dart
final originalVideoUrl = 'http://cbsnewshd-lh.akamaihd.net/i/CBSNHD_7@199302/master.m3u8';
final proxiedUrl = await Peer5.getStreamUrl(originalVideoUrl);

// use the url passing through Peer5's local server
final videoController = VideoPlayerController.network(proxiedUrl);
```

#### Full example:

```dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:peer5/peer5.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _streamUrl;

  @override
  void initState() {
    super.initState();
    Peer5.init();
  }

  @override
  void dispose() {
    super.dispose();
  }

  Future<void> _getStreamUrl() async {
    _streamUrl = await Peer5.getStreamUrl(
        'http://cbsnewshd-lh.akamaihd.net/i/CBSNHD_7@199302/master.m3u8',
    );
    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('Video Player')),
        body: Container(
          padding: const EdgeInsets.all(10.0),
          height: 300.0,
          child: Column(
            mainAxisSize: MainAxisSize.max,
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              RaisedButton(
                onPressed: _getStreamUrl,
                child: const Text('GET STREAM URL'),
                color: Theme.of(context).primaryColor,
                textColor: Colors.white,
                elevation: 30.0,
              ),
              Divider(),
              if (_streamUrl != null) Text(_streamUrl),
            ],
          ),
        ),
      ),
    );
  }
}
```

*That’s all that’s needed for you to get started!*

For more information, read the official documentation:

(iOS) https://docs.peer5.com/platforms/ios/

(Android) https://docs.peer5.com/platforms/android/

## Contributing

Please read [CONTRIBUTING.md](https://gist.github.com/PurpleBooth/b24679402957c63ec426) for details on our code of conduct, and the process for submitting pull requests to us.

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

CHANGELOG.md
## 2.0.0-dev.1

* Null-safety.

## 1.0.0+1

* Improved `pubspec.yaml` parameters.

## 1.0.0

* Implemented the `Peer5.init()` and `Peer5.getStreamUrl()` methods on both iOS and Android.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:peer5_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/peer5.dart
import 'dart:async';
import 'dart:io';
import 'package:flutter/services.dart';

/// Peer5 operates as a service that orchestrates peers and provides analytics
/// and control. The service is activated using our Javascript API in the
/// browser. The client side consists of custom WebRTC p2p logic, an HTTP client
/// and a delivery manager that optimizes speed using the two methods
/// (P2P and HTTP).
///
/// The client connects to multiple peers and to the HTTP server simultaneously
/// and ensures chunks are received in time for playback.
class Peer5 {
  static const MethodChannel _channel = MethodChannel('dev.ianko/peer5');

  /// Check if the `init` method has already been called.
  static bool get initialized => _initialized;
  static bool _initialized = false;

  /// `iOS` Only. Initialize the Peer5 local server.
  static Future<void> init() async {
    if (Platform.isAndroid) {
      _initialized = true;
      return;
    }

    if (initialized) {
      throw 'Peer5 already initialized';
    }

    _initialized = true;

    return _channel.invokeMethod('init');
  }

  /// Get a proxied `url` through Peer5 in exchange for the `originalUrl`.
  /// You will use this `url` to your player.
  static Future<String?> getStreamUrl(String originalUrl) async {
    if (Platform.isIOS && !initialized) {
      throw 'Peer5 needs to be initialized first.';
    }

    return _channel.invokeMethod('streamUrl', <String, dynamic>{
      'url': originalUrl,
    });
  }
}


example/lib/video_player.dart
import 'package:flutter/material.dart';
import 'package:video_player/video_player.dart';

/// Demo video player.
class VideoPlayerDemo extends StatefulWidget {
  /// constructor
  const VideoPlayerDemo({Key? key, required this.url}) : super(key: key);

  /// the video url
  final String? url;

  @override
  _VideoPlayerDemoState createState() => _VideoPlayerDemoState();
}

class _VideoPlayerDemoState extends State<VideoPlayerDemo> {
  late final VideoPlayerController _controller;

  @override
  void initState() {
    super.initState();
    print('Attempting to play: ${widget.url}');
    _controller = VideoPlayerController.network(widget.url!);

    _controller.addListener(() {
      setState(() {});
    });

    _controller.setLooping(true);
    _controller.initialize();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.all(20),
      child: AspectRatio(
        aspectRatio: _controller.value.aspectRatio,
        child: Stack(
          alignment: Alignment.bottomCenter,
          children: <Widget>[
            VideoPlayer(_controller),
            ClosedCaption(text: _controller.value.caption.text),
            _PlayPauseOverlay(controller: _controller),
            VideoProgressIndicator(_controller, allowScrubbing: true),
          ],
        ),
      ),
    );
  }
}

class _PlayPauseOverlay extends StatelessWidget {
  const _PlayPauseOverlay({Key? key, required this.controller}) : super(key: key);

  final VideoPlayerController controller;

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: <Widget>[
        AnimatedSwitcher(
          duration: Duration(milliseconds: 50),
          reverseDuration: Duration(milliseconds: 200),
          child: controller.value.isPlaying
              ? SizedBox.shrink()
              : Container(
                  color: Colors.black26,
                  child: Center(
                    child: Icon(
                      Icons.play_arrow,
                      color: Colors.white,
                      size: 100.0,
                    ),
                  ),
                ),
        ),
        GestureDetector(
          onTap: () {
            controller.value.isPlaying ? controller.pause() : controller.play();
          },
        ),
      ],
    );
  }
}


ios/Classes/SwiftPeer5Plugin.swift
import Flutter
import UIKit
import Peer5Kit

public class SwiftPeer5Plugin: NSObject, FlutterPlugin {
  private var sdk: Peer5Sdk?;

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "dev.ianko/peer5",
      binaryMessenger: registrar.messenger())

    let instance = SwiftPeer5Plugin()

    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "init":
      result(initialize())
    case "streamUrl":
      let url: String? = (call.arguments as? [String: Any])?["url"] as? String
      result(streamURL(url))
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func initialize() -> Any? {
    /// sdk was already initialized
    if (sdk != nil) {
      return nil
    }

    /// get the token from the Info.plist
    let token = Bundle.main.object(forInfoDictionaryKey: "Peer5ApiKey") as? String

    if (token == nil) {
      return FlutterError(
        code: "argument_error",
        message: "Peer5ApiKey not found in the Info.plist",
        details: "Please refer to the Flutter Peer5 plugin documentation on how to add the Peer5ApiKey to your app's Info.plist")
    }

    /// init the local server
    sdk = Peer5Sdk.init(token: token!)
    return nil
  }

  private func streamURL(_ urlString: String?) -> Any? {
    if (sdk == nil) {
      return FlutterError(
        code: "initialization_error",
        message: "Peer5 server was not initialized",
        details: "You need to call Peer5.init() before calling the getStreamUrl() function.")
    }

    guard let unwrappedURLString = urlString,
      let url = URL.init(string: unwrappedURLString)
    else {
      return invalidURLError(urlString)
    }

    let streamURL = sdk?.streamURL(forURL: url)
    return streamURL?.absoluteString
  }
}

/// Returns an error for the case where a URL string can't be parsed as a URL.
private func invalidURLError(_ url: String?) -> FlutterError {
  return FlutterError(
    code: "argument_error",
    message: "Unable to parse URL",
    details: "Provided URL: \(String(describing: url))")
}


android/src/main/kotlin/dev/ianko/peer5/Peer5Plugin.kt
package dev.ianko.peer5

import androidx.annotation.NonNull
import com.peer5.sdk.Peer5Sdk
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

/** Peer5Plugin */
public class Peer5Plugin: FlutterPlugin, MethodCallHandler {
  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "dev.ianko/peer5")
    channel.setMethodCallHandler(Peer5Plugin());
  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  companion object {
    @JvmStatic
    fun registerWith(registrar: Registrar) {
      val channel = MethodChannel(registrar.messenger(), "dev.ianko/peer5")
      channel.setMethodCallHandler(Peer5Plugin())
    }
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "streamUrl") {
      val peer5Url: String = Peer5Sdk.getPeer5StreamUrl(call.argument("url"))
      result.success(peer5Url)
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
  }
}


example/lib/main.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:peer5/peer5.dart';
import 'package:peer5_example/video_player.dart';

void main() => runApp(MyApp());

/// Main example widget
class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final _controller = TextEditingController(text: '');
  String? _streamUrl;

  @override
  void initState() {
    super.initState();
    _initPeer5();
    _controller.addListener(_onTextChange);
  }

  @override
  void dispose() {
    super.dispose();
    _controller.removeListener(_onTextChange);
    _controller.dispose();
  }

  void _onTextChange() {
    setState(() {});
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> _initPeer5() async {
    await Peer5.init();
    print('Peer5 initialized: ${Peer5.initialized}');
  }

  Future<void> _getStreamUrl() async {
    if (_controller.text.isNotEmpty) {
      _streamUrl = await Peer5.getStreamUrl(_controller.text);
      setState(() {});
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('Video Player')),
        body: Container(
          padding: const EdgeInsets.all(10.0),
          child: Column(
            mainAxisSize: MainAxisSize.max,
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            children: <Widget>[
              Row(
                mainAxisSize: MainAxisSize.max,
                crossAxisAlignment: CrossAxisAlignment.end,
                children: <Widget>[
                  Expanded(
                    child: TextField(
                      maxLines: 1,
                      autocorrect: false,
                      cursorWidth: 10.0,
                      keyboardType: TextInputType.url,
                      controller: _controller,
                      textInputAction: TextInputAction.go,
                      decoration: InputDecoration(labelText: 'Video URL'),
                    ),
                  ),
                  InkWell(
                    onTap: _controller.clear,
                    child: Ink(
                      padding: const EdgeInsets.all(6.0),
                      child: Icon(
                        Icons.delete_forever,
                        size: 25.0,
                        color: Theme.of(context).primaryColor,
                      ),
                    ),
                  ),
                ],
              ),
              ElevatedButton(
                onPressed: _controller.text.isEmpty ? null : _getStreamUrl,
                child: const Text('PLAY VIDEO'),
                style: ElevatedButton.styleFrom(
                  elevation: 30.0,
                  textStyle: TextStyle(color: Colors.white),
                  primary: Theme.of(context).primaryColor,
                ),
              ),
              const Divider(),
              if (_streamUrl != null) Expanded(child: VideoPlayerDemo(url: _streamUrl)),
            ],
          ),
        ),
      ),
    );
  }
}
